### PR TITLE
impl `Default` for `HashMap`/`HashSet` iterators that don't already have it

### PR DESCRIPTION
**Pull request by clarfonthey**
This is a follow-up to #128261 that isn't included in that PR because it depends on:

* [x] rust-lang/hashbrown#542 (`Default`)
* [x] `hashbrown` release containing above

It also wasn't included in #128261 initially and should have its own FCP, since these are also insta-stable.

Changes added:

* `Default for hash_map::{Iter, IterMut, IntoIter, IntoKeys, IntoValues, Keys, Values, ValuesMut}`
* `Default for hash_set::{Iter, IntoIter}`

Changes that were added before FCP, but are being deferred to later:

* `Clone for hash_map::{IntoIter, IntoKeys, IntoValues} where K: Clone, V: Clone`
* `Clone for hash_set::IntoIter where K: Clone`

### CODE REVIEW

**Comment by tgross35** [comment]
This seems reasonable but will need an FCP like https://github.com/rust-lang/rust/pull/128261.

r? libs-api

File: library/std/src/collections/hash/set.rs
Comment (dtolnay) [inline library/std/src/collections/hash/set.rs:1556]:
```suggestion
impl<K> Clone for IntoIter<K>
where
    K: Clone,
{
```

**Review by dtolnay — CHANGES_REQUESTED** [review]

**Comment by dtolnay** [comment]
@rust-lang/libs-api:
@rfcbot fcp merge

* `Default for hash_map::{Iter, IterMut, IntoIter, IntoKeys, IntoValues, Keys, Values, ValuesMut}`
* `Clone for hash_map::{IntoIter, IntoKeys, IntoValues} where K: Clone, V: Clone`
* `Default for hash_set::{Iter, IntoIter}`
* `Clone for hash_set::IntoIter where T: Clone`

These are impls of a stable trait for stable types, and therefore instantly stable.

File: library/std/src/collections/hash/set.rs
Comment (clarfonthey) [reply library/std/src/collections/hash/set.rs:1556]: Nice catch. I was mostly holding off fixing the errors before rust-lang/hashbrown#543 is merged, but I had missed this one.

**Comment by clarfonthey** [comment]
May be unprecedented with the FCP still going on, but I'm thinking of dropping the `Clone` impls from this particular PR and deferring them to a later change.

Main issue is that the `Clone` impls are turning out to be substantially more complicated than I thought they'd be, and I think it'd be better to do a bit of refactoring in `hashbrown` to have safer internal parts rather than just hacking together a solution to get these impls to work. It doesn't really have any internal sense of ownership, and so `IntoIter` and `Iter` just reuse the same code, plus an extra pointer stored in the struct to make sure it gets freed.

This also means that we wouldn't depend on the second PR to hashbrown, and only on a new release being pushed out with the `Default` impls. (Which still is necessary, but doesn't require additional review.)

**Review by dtolnay — APPROVED** [review]

**Comment by dtolnay** [comment]
https://github.com/rust-lang/rust/pull/128711#issuecomment-2295044920 &mdash; is this waiting on a hashbrown update?

**Comment by clarfonthey** [comment]
New version of hashbrown was finally pushed, so, rebasing and updating this PR.

@rustbot author

**Comment by clarfonthey** [comment]
Right, I forgot that this API needed to be fixed. I'll mark this as blocked until that's merged, then rebase again.

@rustbot blocked

**Comment by clarfonthey** [comment]
Since I'm working on the other API changes from `hashbrown` 0.15 and this already passed FCP anyway, I'm going to just include this change in the PR that does that. So, closing this, but will link to clarify that FCP passed.

**Comment by clarfonthey** [comment]
Just kidding, there was another PR that added these changes that's already in the queue. So, I'll rebase after that.

**Review by dtolnay — APPROVED** [review]
Thanks!

**Pull request merged**
